onnxruntime/core/optimizer/free_dim_override_transformer.cc

```diff
@@ -67,17 +68,24 @@ Status FreeDimensionOverrideTransformer::ApplyImpl(Graph& graph, bool& modified,
           return Status(ONNXRUNTIME, INVALID_ARGUMENT, "Invalid free dimension override.");
         }
 
+        if (dimension.has_dim_value() && !dimension.has_dim_param()) {
```

Comment:
Dimension can have 3 states.
has_dim_value
has_dim_param
neither

Maybe check !dimension.has_dim_param() only so the other two states result in continuation.

Reply:
I was going by what happens in the remainder of the loop: it sets dim_value and clears dim_param, so if this condition is true, the remainder of the loop is a no-op. From your description, looks like `!dimension.has_dim_param()` check is redundant and we can remove it. However, if we remove `dimension.has_dim_value()` instead, this will miss the "neither" case. Did I misunderstand this?

Reply:
You possibly missed the 'not' part in my suggestion to only check has_dim_param.

If there's no dim_param, there's nothing the loop can do is there, so you could probably do that check as soon as you initialize 'dimension' at the start of the loop?

Reply:
The override is based on denotation, not dim_param. Is it not possible that the dimension has a denotation, but neither dim_value, nor dim_param, and in that case we still want to set dim_value?